### PR TITLE
lib: fix parsing of amounts with a whitespace digits separator 

### PR DESCRIPTION
**Pull request by awjchen**
This PR addresses a problem which was found in issue #749. This PR does not address that issue.

The problem was that we fail to parse the following journal
```
commodity 1 000  USD

2018-01-01
  (a)   USD1 000
```
failing with the following error
```
hledger: test.journal:1:12:
unexpected space
expecting digit
```
We fail to parse this next journal as well
```
2018-01-01
  (a)   USD1 000 ; comment
```
```
hledger: test2.journal:2:9:
unexpected 'U'
expecting '=', '{', end of input, or newline
```
It seems that the issue is caused by the use of a space as the digits separator and trailing whitespace after the amount. The current parser consumes the trailing whitespace as a separator and then expects at least another digit, which it does not find. This patch groups together the separator and the following digit group. With this patch, the above two journals parse successfully.

### CODE REVIEW

**Comment by ehildenb** [comment]
Can you add a test or two for this case? Perhaps the above examples.

**Comment by simonmichael** [comment]
I added some functional tests for #749 in tests/journal/amounts-and-commodities.test. You can just uncomment them and `make functest` or `COLUMNS=80 shelltest --execdir --hide-successes -j8 tests/journal/amounts-and-commodities.test [-i...]`.

File: hledger-lib/Hledger/Read/Common.hs [hunk 1]
```diff
@@ -763,8 +763,9 @@ rawnumberp = do
         leadingDigits <- some digitChar
         option (Nothing, [leadingDigits]) . try $ do
             firstSep <- oneOf sepChars <|> whitespaceChar
-            groups <- some digitChar `sepBy1` char firstSep
-            return (Just firstSep, leadingDigits : groups)
+            secondGroup <- some digitChar
+            otherGroups <- many $ try $ char firstSep *> some digitChar
```
Comment (ony) [inline hledger-lib/Hledger/Read/Common.hs]: Hmm... Interesting how this is different from previous code that suppose to produce list of one or more elements with non-zero list of digit characters separated by `firstSep`? Am I missing something?

Comment (awjchen) [reply hledger-lib/Hledger/Read/Common.hs]: Yes, the new code also parses a list of one or more numbers separated by `firstSep`, where each number is a non-zero list of digit characters. The difference is only that I have added `try` so that a separator and the following digits are parsed together as an atomic unit. The old code failed to parse the above examples because it interpreted trailing whitespace as a separator.  With the new code, whitespace will only be interpreted as a separator if digits follow it.

**Comment by awjchen** [comment]
I've uncommented tests 11 and 12 in `tests/journal/amounts-and-commodities.test` and updated this PR.

Test 10 is also relevant, but as it is, it does not yet pass since we have not fully resolved issue 749.

File: hledger-lib/Hledger/Read/Common.hs (hunk 1 above)
Comment (ony) [reply hledger-lib/Hledger/Read/Common.hs]: Hmm... Yep. Confirmed with test-case

```
### Failure in /home/nikolay/contrib/hledger/hledger-lib/Hledger/Read/Common.hs:757: expression `parseTest rawnumberp "1 000 USD"'
expected: RawNumber (Just ' ') ["1","000"] Nothing
 but got: 1:2:
          unexpected space
          expecting digit
```

Apparently `sepBy` in `megaparsec` expects that if separator were parsed successfully then next entry is guaranteed to parse correctly.

Though I still cannot understand error message. `notFollowedBy $  (whitespaceChar >> digitChar)` should say something about unexpected space followed by digit.

Comment (awjchen) [reply hledger-lib/Hledger/Read/Common.hs]: It seems that the problem is that the remaining text that the `notFollowedBy ...` line is given to parse is ` 000 USD`, which triggers its second alternative. The first section of `rawnumberp` doesn't consume the ` 000` due to `try` in `option (...) . try $ do` that encloses the failure on the `groups <- some digitChar ...` line.

**Comment by awjchen** [comment]
Having the approval of @ony (and myself), I will merge this PR.

**Comment by simonmichael** [comment]
Thanks @awjchen for the fix and @ony for the review. The results look good. 

In general, use whichever kind of test works best for you, we don't have a firm policy. But it's a good point about doctests being part of the API documentation. hunittests run fastest but are a chore to write, shell tests run pretty fast and are easy to write and understand, doctests are easy to write and provide useful documentation but alas run extremely slowly. Please feel free to add notes to https://github.com/simonmichael/hledger/wiki/Tests.

**Comment by simonmichael** [comment]
easytests are new and might be easier than hunittests.

**Comment by ony** [comment]
@awjchen , doctests are simple unit-tests for developers that helps to catch errors that propagates further. Especially useful for parsers. They also help to understand some corner cases of API.

In this particular case issue were isolated because of extra check `notFollowedBy` in the end. But if that line would be missing you'll get error in a next parser or will get wrong interpretation.

**Comment by awjchen** [comment]
Ah, I haven't before considered whether issues were propgated or isolated. Thanks for the tips!
